### ISSUE TITLE
High CPU usage

### ISSUE THREAD

**Issue by ammgws** [greshake/i3status-rust#569]
Seeing high usage again, though the cause may be different to #269 

CPU usage:
![20200219_19h08m11s](https://user-images.githubusercontent.com/20397027/74824584-ba366100-534b-11ea-8aea-da19992f50b6.png)

Blocks in use: pacman, custom, music, disk_space, memory, cpu, load, net, sound, ibus, weather, time, focused_window
![image](https://user-images.githubusercontent.com/20397027/74824609-c6222300-534b-11ea-98bc-e42fb58d0383.png)

Call trace log:
![image](https://user-images.githubusercontent.com/20397027/74824631-cde1c780-534b-11ea-83d1-f18471832566.png)

This is what the wait4 calls look like but I'm not sure they are the issue:
![image](https://user-images.githubusercontent.com/20397027/74828215-100e0780-5352-11ea-85d0-5ca50a672e04.png)

Read errors, not sure what is causing these
![image](https://user-images.githubusercontent.com/20397027/74828762-0507a700-5353-11ea-90e9-c4b673fd3605.png)

Need to find some other ways to attack this but might be a long while until this happens again since I need to reboot the computer now. Posting in case people have some ideas.

**Comment by ammgws** [comment]
This might be related to [an issue](https://github.com/altdesktop/i3ipc-python/issues/147) I am having with the sway IPC/script (was using the `focused_window` block at the time. Going to close this until it happens again (not using the block anymore)

**Comment by xxxserxxx** [comment]
Can we re-open this?  It's a persistent issue on my machine; i3status-rs ranges between 12% and 20% constant CPU use; i3bar is at 30%.  When running different status programs, i3bar does not demonstrate this high use.  I'm happy to run debugging.

After I wrote all of this out, I ran i3status-rs on the command line by itself; it echo'd to the STDOUT as expected, paused for a second, and then started dumping out non-stop.  So, that's where the CPU use is coming from.  I'm going to try shutting down widgets one at a time to see if one of them is causing i3status-rs to refresh without pause.

### Configuration information
* i3status-rust from git checkout (0.13.0 // 5e6e1ed4e47 @ 2020-04-30)
* i3-gaps 4.18.1
* Linux 5.6.8-arch1-1 #1 SMP PREEMPT x86_64

### Blocks enabled 
If no interval listed, I'm using the default
- weather (openweathermap)
- music (mpd)
- disk_space, 20
- memory
- battery, 10
- xrandr
- net (wifi, up/down & graph all disabled), 5
- net (eth, up/down & graph disabled), 5
- notmuch (I'm on Arch but I built i3status-rs explicitly for this module)
- pacman
- toggle (TimeWarrior `timew get dom.active`), 12
- load, 1
- backlight
- sound
- keyboard_layout (localebus)
- time, 60

## Comparative CPU use
I ran several different i3status replacements (all under i3bar); each had at least 30s to settle after changing the config and reloading i3:

#### Run 1
```
%CPU %MEM CMD
30.9  0.3 i3bar --bar_id=bar-0 --socket=/run/user/1001/i3/ipc-socket.1441
13.4  0.3 /home/ser/.local/bin/i3status-rs /home/ser/.i3/rust.toml
```
#### Run 2
Restart i3; 30s settle
```
%CPU %MEM CMD
33.3  0.3 i3bar --bar_id=bar-0 --socket=/run/user/1001/i3/ipc-socket.1441
14.3  0.3 /home/ser/.local/bin/i3status-rs /home/ser/.i3/rust.toml
```
#### Bumblebee 
Widgets: python; weather, bluetooth, batt, mem, load, nic, brightness, pulseaudio, cryptocoin, keyboard, datetime
```
%CPU %MEM CMD
 0.7  0.3 i3bar --bar_id=bar-0 --socket=/run/user/1001/i3/ipc-socket.1441
 1.8  0.4 python /usr/bin/bumblebee-status --markup pango -m weather bluetooth battery
```
#### goblocks
Widgets: disk, load, mem, temp, wifi, battery, volume, time
```
%CPU %MEM CMD
 1.4  0.2 i3bar --bar_id=bar-0 --socket=/run/user/1001/i3/ipc-socket.1441
 1.3  0.1 /usr/bin/goblocks
```
#### Back to i3status-rs
```
%CPU %MEM CMD
29.7  0.3 i3bar --bar_id=bar-0 --socket=/run/user/1001/i3/ipc-socket.1441
13.7  0.3 /home/ser/.local/bin/i3status-rs /home/ser/.i3/rust.toml
```

**Comment by xxxserxxx** [comment]
Well, that was easy.  I haven't vetted *all* of the blocks yet, but it's music.  If that's enabled, i3status-rs just dumps out nonstop with no pause.  I tried both with specifying `player`, and not specifying `player`.  My system uses mpd-mpris.

One thing I've noticed is that i3status-rs doesn't appear to buffer updates; whatever threading is going on to support different intervals, if two arrive close to each other, i3status-rs will dump both out one immediately after the other.  Since every message is the *entire* status line, this causes "throw-away" updates which are displayed for so short a time they're not visible. The more blocks that are enabled, the worse this becomes, with sometimes several updates coming effectively (for the user) at the same time.

However, disabling only the music block greatly improves the CPU use and brings it down below goblocks:
```
%CPU %MEM CMD
 2.4  0.3 i3bar --bar_id=bar-0 --socket=/run/user/1001/i3/ipc-socket.1441
 0.9  0.3 /home/ser/.local/bin/i3status-rs /home/ser/.i3/rust.toml
```
It's still on par with goblocks total, but only because it's causing i3bar to work far harder than the others with the unnecessary IPC messages.
